[영화 극장 선택] 해나 1, 2단계 미션 제출합니다.

hxeyexn opened the pull request:
안녕하세요! 페로로 😊

지난 Lv1 블랙잭 미션 때 리뷰이로 매칭되었었는데 그때 정말 많이 배웠습니다!! 이번 미션도 잘 부탁드리겠습니다 🙇
[2cf5c4c](https://github.com/woowacourse/android-movie-theater/pull/55/commits/2cf5c4c185e98b9fab7eccccfcf5f10005cc5f05) 해당 커밋은 이전 미션 마이그레이션 커밋입니다!

[시현영상 링크](https://drive.google.com/file/d/1bTg-HeqAXAO0-KjXSyBaVqxMjwg2zYuk/view?usp=sharing) 첨부합니다!

<br>

이번 미션의 질문 사항은 아래 작성해두었습니다!

### 질문
- 데이터 바인딩 시 프레젠터 안에 있는 모델을 private를 풀어야 하는데, 이런 방식으로 바인딩을 진행하면 안전성이 떨어지는 것인가요? 바인딩을 처음 사용해봐서 궁금증이 생겼습니다.
- ex) presenter에 private로 선언된 private val number = 0을 데이터바인딩을 위해서 val number = 0으로 한 후, 액티비티에 있는 private val presenter를 val presenter로 수정 -> xml에서 presenter.number를 사용한 경우

hxeyexn committed:
init: 영화 예매 미션 마이그레이션

hxeyexn committed:
refactor: 도메인 추가를 위해 데이터베이스 수정

hxeyexn committed:
chore: data binding 의존성 추가

hxeyexn committed:
chore: fragment 의존성 추가

hxeyexn committed:
docs: 기능 명세 작성

hxeyexn committed:
fix: 데이터베이스 변경에 따른 테스트 코드 수정

hxeyexn committed:
refactor: 홈 화면 activity를 fragment로 변경

hxeyexn committed:
refactor: Movie의 screeningTimes 제거하도록 도메인 수정

hxeyexn committed:
feat: Theater 도메인 모델 생성

hxeyexn committed:
feat: Theater 데이터베이스 생성과 테스트 작성

hxeyexn committed:
fix: Movie 도메인 모델 수정에 따른 테스트코드 오류 해결

hxeyexn committed:
feat: 극장 선택 Contract 생성과 테스트 코드 작성

hxeyexn committed:
design: 극장 선택 ui 구현

hxeyexn committed:
feat: 극장 아이템 리싸이클러뷰 어댑터 구현

hxeyexn committed:
feat: TheaterSelectionFragment MVP 패턴 적용

hxeyexn committed:
feat: Home 및 극장 선택 화면 이동 기능 구현

hxeyexn committed:
feat: 화면 전환 시 movieId 전달 기능 구현

hxeyexn committed:
feat: HeadCount 객체 구현

- Ticket의 count를 HeadCount라는 객체로 분리

hxeyexn committed:
chore: default activity 수정

hxeyexn committed:
design: bottom icon 추가

hxeyexn committed:
design: bottom navigation view 구성

hxeyexn committed:
feat: bottom navigation 연결

hxeyexn committed:
refactor: 티켓 분리

hxeyexn committed:
refactor: home 화면 databinding 추가

hxeyexn committed:
style: 린트 체크 & 코드 포맷 수정

hxeyexn committed:
feat: 극장 선택 기능 구현

hxeyexn committed:
feat: movieId에 맞는 극장 데이터 찾아오기

hxeyexn committed:
refactor: 예약 상세화면 인원 수 증감 버튼 data binding 구현

hxeyexn committed:
feat: 극장에 맞는 상영시간 불러오는 기능 구현

hxeyexn committed:
refactor: 좌석 선택 화면 이동 버튼 data binding 구현

hxeyexn committed:
fix: 상영시간 스피너 인덱스 오류 해결

hxeyexn committed:
feat: 예매 완료 시 ticket이 발행되도록 변경

hxeyexn committed:
feat(ReservationFinshed): 티켓 발행되지 않았을 때 오류 처리

hxeyexn committed:
refactor: ui 테스트 패키지 이동

- db 패키지로 잘못 들어가 있어 db 패키지 밖으로 이동 시킴

hxeyexn committed:
design: fragmentContainer 하단이 가리는 현상 해결

hxeyexn committed:
refactor: 영화 및 광고 목록 item 갯수 count 로직 변경

hxeyexn committed:
refactor: 홈 화면 -> 극장 선택 화면으로 movieId 전달 로직 수정

- bundle로 넘겨주도록 수정

hxeyexn committed:
feat: 극장 db에 더미데이터 추가

hxeyexn committed:
test: 극장 선택 UI 테스트 코드 추가

hxeyexn committed:
style: 코드 포맷팅

hxeyexn committed:
design: 극장 이름 textview 추가

hxeyexn committed:
feat: 극장 선택 데이터 바인딩 기능 구현

hxeyexn committed:
refactor: 예약 상세 ui 수정에 따른 테스트코드 리팩토링

hxeyexn committed:
test: 예약 확인 presenter에 극장 이름을 보여주는 테스트 추가

hxeyexn committed:
refactor: 프레젠터 위치에 맞게 테스트 파일 패키지 이동

hxeyexn committed:
chore: parentActivity 스택 수정

hxeyexn committed:
refactor: view 변경에 따른 Home 테스트 수정

- ReservationHomeActivity를 HomeFragment로 교체

hxeyexn committed:
refactor: 패키지 이동 및 정리

hxeyexn committed:
docs: 기능 명세서 정리

hxeyexn committed:
style: 코드 포맷팅 및 상수 정리

hxeyexn committed:
refactor: binding 선언 방법 변경

laco-dev commented on app/src/androidTest/java/woowacourse/movie/view/home/HomeFragmentTest.kt:
```diff
@@ -7,35 +7,45 @@ import androidx.test.espresso.assertion.ViewAssertions.matches
+class HomeFragmentTest {
     @get:Rule
-    val activityRule = ActivityScenarioRule(ReservationHomeActivity::class.java)
+    val activityRule = ActivityScenarioRule(MainActivity::class.java)
```
> 따라서 Fragment는 Activity나 다른 Fragment와 의존성이 없어야 한다.

프래그먼트가 어떠한 액티비티에 의존하면안된다고하네요!
만약, 이 프래그먼트가 MainActivity가 아닌 다른 액티비티에도 여러곳에서 사용이 된다면 어떨까요?

hxeyexn replied:
프래그먼트 UI 테스트에 대한 숙지가 덜 되었던 것 같습니다..!
`launchFragmentInContainer`를 이용하는 방법으로 화면 테스트하도록 변경했습니다 😊
[9fd1dd0](https://github.com/woowacourse/android-movie-theater/pull/55/commits/9fd1dd0ebfe467d6b8c9fbb0c1538b56c15983e1)

laco-dev commented on app/src/androidTest/java/woowacourse/movie/view/theater/TheaterSelectionFragmentTest.kt:
```diff
@@ -0,0 +1,83 @@
+
+    @Test
+    fun `극장_목록을_보여준다`() {
+        Thread.sleep(1000)
```
화면을 테스트하는 다른 방법은 없었을까요?
모든 테스트마다 1초씩 멈추게 된다면, 화면이 수십개 수백개라면 테스트에 수십분 이상이 걸리지 않을까요?

hxeyexn replied:
위 코멘트와 동일하게 `launchFragmentInContainer` 를 이용해 화면 테스트하도록 변경했습니다! [07a6881](https://github.com/woowacourse/android-movie-theater/pull/55/commits/07a688160a9dc34728a92d08e6a970d83713a9f3)

laco-dev commented on app/src/main/java/woowacourse/movie/model/theater/Theater.kt:
```diff
@@ -0,0 +1,10 @@
+    val theaterId: Int,
+    val name: String,
+    val screeningTimes: ScreeningTimes,
+    val movieId: Int,
```
하나의 극장에서 여러개의 영화를 보여주어야 한다면 어떨까요??

극장에 상영시간들과 영화를 고정하는 대신, 현실 세계의 극장으로 비유해보면 좋을 것 같아요.
```
극장 A
  - 09:00 해리포터
  - 10:00 어벤저스
  - 11:00 우테코라이브
```

hxeyexn replied:
페어와 미션을 진행하면서 어색하다고 느꼈던 부분인데 조언 주셔서 감사드립니다! 🙇
피드백 주신 내용 참고하여 아래와 같은 구조로 변경했습니다!

``` kotlin
val theaters: List<Theater> =
    listOf(
        Theater(
            theaterId = 0,
            name = "선릉 극장",
            seolleungScreeningSchedule,
        ),
        ... 
    )
```

```kotlin
private val seolleungScreeningSchedule =
    listOf(
        ScreeningSlot(
            screeningTime = LocalTime.of(9, 0),
            movieId = 0,
        ),
        ScreeningSlot(
            screeningTime = LocalTime.of(10, 0),
            movieId = 0,
        ),
        ScreeningSlot(
            screeningTime = LocalTime.of(11, 0),
            movieId = 1,
        ),
        ScreeningSlot(
            screeningTime = LocalTime.of(12, 0),
            movieId = 2,
        ),
        ...
    )
```

laco-dev replied:
`Screening`라고 짓지 않은 이유는 어색해서였을까요?

hxeyexn replied:
`Screening`이 한 타임의 상영이라는 것보단 여러 개의 상영을 뜻하는 느낌이 들어서 ScreeningSlot이라고 네이밍 했습니다!
`ScreeningSchedule` 역시 `Screening`으로 지었을 때 상영시간표라는 걸 명시적으로 표현하지 못한다고 생각했습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/model/ticket/Ticket.kt:
```diff
@@ -1,49 +1,13 @@
+data class Ticket(
+    val movieId: Int,
+    val theaterId: Int,
+    val seats: Seats,
+    val screeningDateTime: ScreeningDateTime,
+    val amount: Int,
```
seats와 amount는 어떤 관계를 가지나요?
좌석이 5개인데 amount는 2개가 될 수 있는 개념일까요?

hxeyexn replied:
좌석들의 총 금액(amount) 이라는 관계를 생각했습니다..! 

질문 주신 부분에 대해서 두 가지를 고민 해봤는 데, 제가 질문의 의도를 잘 파악한건 지 확신이 서지 않네요 ㅠㅠ

> 1. 변수명이 동료 개발자가 봤을 때 좌석들의 총 금액이라는 걸 쉽게 인지하기 어렵다. -> amount를 좀 더 의미있는 변수명으로 바꾼다. ex) `extendedPrice` 등
> 2. 티켓이 amount를 가지지 않고 finished 화면에서 ticket의 seats를 이용해 seats.calculateAmount()를 불러 가격을 보여준다.
> 

질문 주신 내용이 1, 2 중 어느 쪽에 더 가까운 지, 아니면 제가 아예 다른 방향으로 잘못 생각하고 있는 건지에 대해 가이드를 좀 더 얻을 수 있을까요? 🙇

laco-dev replied:
1차적으로는 수량으로 느껴졌어요. 그래서 좌석 수 == 수량 동일하다고 생각을 했는데요.
가격의 입장으로 보면 이해가 되기도 합니다.

가격은 티켓이 가져야 할 것이 맞다고 봐요. 다만 가격이 현재 좌석과 의존해서 결정되고 있어서 어떻게 의존 관계를 설정하는가 차이로 보여요.

```kotlin
class Ticket(seats: Seats) {
   fun getTotalAmount(): Int { seats.xxx() }
}
```

다만 티켓은 모든 가격이 결정되고 난 다음의 영수증이라고 본다면 가격은 티켓을 발급하면서 생성될 수도 있다고 봐요.
결과적으로 해석하기에 따라 지금 구조도 문제는 없어 보여서 의견만 드렸습니다.

hxeyexn replied:
감사합니다! 위와 같은 구조가 좀 더 책임과 역할이 잘 들어나는 것 같아서 피드백 참고하여 변경했습니다~

laco-dev commented on app/src/main/java/woowacourse/movie/model/ticket/HeadCount.kt:
```diff
@@ -0,0 +1,40 @@
+    fun restore(count: Int): HeadCount {
+        restoreCount(count)
+        return this
+    }
```
복구라는 개념을 도메인에 적용할 필요가 있을까요?
동적으로 상태를 관리하는 대신 새로운 HeadCount를 생성하면 되지 않을까요?

```kotlin
val headCount = HeadCount(count)
```

hxeyexn replied:
복구라는 개념을 도메인에 적용할 필요가 전혀 없어보이네요..! 😱 
새로운 HeadCount를 생성하는 방식으로 변경했습니다! [399316a](https://github.com/woowacourse/android-movie-theater/pull/55/commits/399316a52f0f5c5b484ec262a597d9f82c6b7a96)

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/finished/ReservationFinishedContract.kt:
```diff
@@ -1,25 +1,24 @@
-            ticket: Ticket,
-            seats: Seats,
-        )
+        fun loadTicket(ticket: Ticket)
```
Ticket을 불러오는 함수인데 Ticket을 파라미터로 넣는게 정상일까요?

hxeyexn replied:
아래서 조언주신 대로 ticket도 고정값이라고 생각해 Presenter의 생성자로 주입하도록 변경했습니다! [80c39e0](https://github.com/woowacourse/android-movie-theater/pull/55/commits/80c39e0fa1cbc8b86cad639b34ea5d0df7f11c4a)

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/reservation/ReservationDetailContract.kt:
```diff
@@ -34,19 +35,16 @@ interface ReservationDetailContract {
 
         fun loadScreeningTimes(
-            movieId: Int,
+            theaterId: Int,
```
극장과 영화는 영화를 예매하는 과정에는 항상 고정값일텐데요.
프레젠터의 생성자로 받지 않고 함수에 받도록 한 이유가 있을까요?

hxeyexn replied:
극장과 영화가 항상 고정값이라는 것에 대해 미처 생각하지 못해 함수로 받도록 했습니다... 🥹
함수가 아닌 프레젠터의 생성자로 받도록 변경했습니다! [0d2a0a5](https://github.com/woowacourse/android-movie-theater/pull/55/commits/0d2a0a58ff50f8986ad21058b738fa91229b4db7)

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/reservation/ReservationDetailContract.kt:
```diff
@@ -34,19 +35,16 @@ interface ReservationDetailContract {
+        fun initializeReservationButton(dateTime: ScreeningDateTime)
 
-        fun handleNumberOfTicketsBounds(result: ChangeTicketCountResult)
+        fun handleHeadCountBounds(result: ChangeTicketCountResult)
```
수량 변경에 대한 결과를 뷰가 프레젠터에게 호출할 일이 있을까요?

hxeyexn replied:
페로로 말씀대로 수량 변경에 대한 결과를 뷰가 프레젠터에게 호출할 일이 없어보이네요..! 🥲
presenter만 해당 메서드를 가지고 있도록 변경했습니다! [29d4abc](https://github.com/woowacourse/android-movie-theater/pull/55/commits/29d4abc1b5fa81f359beeafda93d409b153c0148)

질문의 의도를 잘 파악한 건지 확신이 서지 않는 데 혹시 제가 코멘트를 잘못 이해하고 있다면 알려주시면 정말 감사하겠습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/reservation/ReservationDetailPresenter.kt:
```diff
@@ -1,60 +1,58 @@
+    private val theaterDao: TheaterDao,
 ) : ReservationDetailContract.Presenter {
-    var ticket = Ticket()
+    var headCount = HeadCount()
```
View에서 필요한 로직은 전부 View.xxx 형태로 호출하고 있는데, 이 상태는 왜 public으로 접근을 할 수 있게 되있는 걸까요?

hxeyexn replied:
해당 상태를 private으로 변경해 외부에서 presenter의 상태에 접근하지 못하도록 수정했습니다! [86425ef](https://github.com/woowacourse/android-movie-theater/pull/55/commits/86425efcbd8c565595dcae99c6439b135f18f817)

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/reservation/ReservationDetailPresenter.kt:
```diff
@@ -1,60 +1,58 @@
-        val result = ticket.increaseCount()
-        handleNumberOfTicketsBounds(result)
+    override fun increaseHeadCount(count: Int) {
+        headCount = HeadCount(count)
```
수량을 더한다라고 생각하면, 현재 가지고 있는 상태에서 더하는 것을 생각할텐데, 이 함수는 단순히 덮어 씌우고 있지 않나요?

hxeyexn replied:
현재 가지고 있는 상태에서 더하도록 수정했습니다! [fc4d22f](https://github.com/woowacourse/android-movie-theater/pull/55/commits/fc4d22f145f242a6f53244f7a2b421ae4b98d585)

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/reservation/ReservationDetailPresenter.kt:
```diff
@@ -1,60 +1,58 @@
-        ticket = Ticket(ticket.count, dateTime)
-        view.navigateToSeatSelection(movieId, ticket)
+    override fun initializeReservationButton(dateTime: ScreeningDateTime) {
+        view.navigateToSeatSelection(dateTime, headCount)
```
프레젠터는 UI와 직접적인 연관이 없도록 작성되어야 합니다.
*모델과 비즈니스 로직을 호출하는 관계를 가지기 때문입니다.

즉, 예약 "버튼"이라는 것과 관련 없이 기능 관점으로 이름이 작성되어야합니다.
예를 들어 프레젠터의 예약 버튼을 초기화 하라는 행동이 왜 네비게이션으로 이어지는지 이해가 안됩니다.

hxeyexn replied:
프레젠터는 UI와 직접적인 연관이 없도록 작성되어야 하는 걸 숙지하지 못했네요 ㅠㅠ 😭 
기능 관점으로 네이밍 변경했습니다! [0e367e1](https://github.com/woowacourse/android-movie-theater/pull/55/commits/0e367e194eb8b71ac8027cbfcff8a8c31156250b)

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/reservation/SeatSelectionContract.kt:
```diff
@@ -31,6 +33,12 @@ interface SeatSelectionContract {
+            movieId: Int,
+            theaterId: Int,
+            screeningDateTime: ScreeningDateTime,
+        )
```
마찬가지로 좌석 선택 프레젠터는 이미 영화 아이디와 극장 아이디를 알고있을거에요.
동적으로 바뀔일이 없으니까요

hxeyexn replied:
Presenter의 주 생성자로 movieId와 theaterId 주입하도록 변경했습니다! [73323bd](https://github.com/woowacourse/android-movie-theater/pull/55/commits/73323bde60bc087be8d6f574d40d4fa8bca0f8b2) , [688417c](https://github.com/woowacourse/android-movie-theater/pull/55/commits/688417c304be7effb002c60d9d58fe1a5fa8fd7e)

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/reservation/SeatSelectionPresenter.kt:
```diff
@@ -3,22 +3,24 @@ package woowacourse.movie.presenter.reservation
-        view.setConfirmButtonEnabled(ticket.count)
-        view.showAmount(ticket.calculateAmount(seats))
+        headCount.restore(count)
+        view.setConfirmButtonEnabled(headCount.count)
```
enabled 를 설정하는데 Int값을 넣나요?

hxeyexn replied:
Boolean 값을 넣어 enabled를 설정하도록 변경했습니다! [513b203](https://github.com/woowacourse/android-movie-theater/pull/55/commits/513b203fac953074ef172202fced17d8ace5c8d9)

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/theater/TheaterSelectionContract.kt:
```diff
@@ -0,0 +1,14 @@
+interface TheaterSelectionContract {
+    interface Presenter {
+        fun loadTheater(theaterId: Int)
+    }
```
극장 목록을 불러와야하는데 극장 아이디를 어떻게 미리 알고있나요? 🤔

hxeyexn replied:
헉 네이밍이 동작을 명확히 나타내지 못하고 있네요 ㅠㅠ 😱 
`sendTheaterInfoToDetail()`이라는 이름으로 변경했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/theater/TheaterSelectionPresenter.kt:
```diff
@@ -0,0 +1,10 @@
+package woowacourse.movie.presenter.theater
```
패키지를 컴포넌트별로 나누면 어떤 문제가 있을까요?
만약에 MVP 아키텍처에서 MVVM 아키텍처로 바뀌면 presenter도 전부 바꿔야 하는걸까요?
혹은 화면이 수백개가 된다면, 수백개 x N개씩 패키지가 늘어나지는 않을까요?

그렇다면 화면별로 패키지를 나눈다면 어떤 차이가 있을까요?

hxeyexn replied:
패키지를 컴포넌트별로 나눈다면 MVP 아키텍처에서 MVVM로 전환하는 경우 대부분의 Presenter를 ViewModel로 대체해야 해서 광범위한 코드 리팩토링이 요구될 것 같습니다. 또한 화면 수가 많아지면 각각의 화면에 해당하는 컴포넌트를 찾는 데 시간 비용이 많이 들겠네요..!

화면별로 패키지를 나눈다면 아키텍처 구조가 변경되더라도 패키지가 크게 변경될 일이 없어서 유지보수성이 좋을 것 같고, 각각의 화면에 해당하는 파일을 찾기가 훨씬 수월할 것 같습니다!

아키텍처 구조가 변경될 수 있다는 점을 고려하지 못했던 것 같습니다 😭 view, presenter, contract 가 한 패키지에 존재하도록 수정했습니다! [2645ecd](https://github.com/woowacourse/android-movie-theater/pull/55/commits/2645ecdf526c244b5b36110a7d424907a2368a80)

laco-dev commented on app/src/main/java/woowacourse/movie/utils/MovieUtils.kt:
```diff
@@ -17,6 +22,8 @@ object MovieUtils {
         return "%s~%s".format(start, end)
     }
 
+    fun convertScreeningInfoFormat(screeningTimeCount: Int): String = "${screeningTimeCount}개의 상영 시간"
```
기능을 만드는데 있어서 유틸리티 라는 것을 사용하는 것은 최대한 지양해야 합니다.
어? 그냥 여러번 쓰이면 다 유틸이네? 라는 위험한 생각은. 돌이킬 수 없는 스파게티가 될 수 있습니다.

특정 상영 갯수로 하여금 문자열을 출력하는 역할은 MovieUtil이라는 객체가 할 일인가요?

laco-dev replied:
추가로, string 리소스를 사용하는 방법을 학습하고 사용했을 거에요.
왜 여기는 하드코딩이 되어있나요?

hxeyexn replied:
`여러번 쓰이면 다 유틸이네?` 라는 생각을 자주 했는 데 반성하게 되네요..🥲
해당 메서드는 DataBinding을 다시 적용하면서 MovieUtil에서 제거했습니다! 

string 리소스로 추출하는 걸 놓친 부분도 적용해두었습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/utils/MovieUtils.kt:
```diff
@@ -51,4 +58,28 @@ object MovieUtils {
+            add(fragmentLayoutResource, nextFragment)
+            addToBackStack(null)
+        }
+    }
```
이 동작들을 모든 화면과 모든 화면에 재사용 할 수 있을 정도로 신뢰하나요?
*다른 개발자들이 이 함수를 써야만 한다는 것을 사전에 잘 인지 할 수 있을지가 걱정됩니다.
하나라도 달라지면 파라미터가 그때마다 늘어나야만 할까요?

hxeyexn replied:
페로로 말씀대로 해당 동작들은 모든 화면에 재사용할 수가 없겠네요..!
해당 동작이 필요한 위치로 이동시켰습니다! [9d7de36](https://github.com/woowacourse/android-movie-theater/pull/55/commits/9d7de367e5a04e94fec9bac6272f8f6524d74ecd)

laco-dev commented on app/src/main/java/woowacourse/movie/view/finished/ReservationFinishedActivity.kt:
```diff
@@ -2,69 +2,79 @@ package woowacourse.movie.view.finished
+        binding.textViewReservationFinishedTicketPrice.text = convertAmountFormat(this, ticket.amount)
+        binding.textViewReservationFinishedSeats.text =
+            seats.joinToString(getString(R.string.reservation_finished_seat_separator)) { "${it.row}${it.column}" }
+        presenter.loadTheater(ticket.theaterId)
```
예약 내역을 보여주는데 다시 극장을 요청하는 것은 왜인가요?

hxeyexn replied:
예약 내역을 보여주는데 다시 극장을 요청하는 게 어색해보이네요..! 🤣
티켓 발급 시에 티켓에 극장 이름이 담기도록 변경했습니다! [36ee039](https://github.com/woowacourse/android-movie-theater/pull/55/commits/36ee039bc3230bea37290246c387dbf8f08e1e52)

laco-dev commented on app/src/main/java/woowacourse/movie/view/finished/ReservationFinishedActivity.kt:
```diff
@@ -2,69 +2,79 @@ package woowacourse.movie.view.finished
+        }.onFailure {
+            showErrorToast()
+            finish()
+        }
```
티켓을 받는 동작이라면 티켓을 반환하는 것이 자연스럽지 않을까요?
티켓이 없다고 화면을 꺼버린다면 2가지 동작을 하는 것이 아닐까 궁금하네요.

혹은 다음 코드가 어떻게 동작할지 찾아보세요.
```kotlin
receiveTicket()
Log("Hello World") // 이 코드는 호출이 될까요?
presenter.foo() // 이 코드도 호출이 될까요?
```

hxeyexn replied:
티켓을 반환하는 방식으로 변경했습니다! [80c39e0](https://github.com/woowacourse/android-movie-theater/pull/55/commits/80c39e0fa1cbc8b86cad639b34ea5d0df7f11c4a)
티켓이 없다고 바로 화면을 꺼버리면 사용자 입장에서 당황스러울 수 있겠네요..! 🥲

티켓을 제대로 전달받지 못했을 때 해당 화면을 유지한 채로 스낵바를 이용해 사용자에게 에러 문구(ex. `예기치 못한 오류가 발생해 티켓이 발급되지 않았습니다. 티켓을 처음부터 다시 발급하시겠습니까?`)를 띄우고, 사용자가 홈으로 이동 버튼을 눌렀을 때만 홈 화면으로 이동해 예매를 다시 진행할 수 있도록 변경했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/view/home/HomeFragment.kt:
```diff
@@ -0,0 +1,72 @@
+        val movieCatalogAdapter =
+            MovieCatalogAdapter(
+                ScreeningDao().findAll(),
+                AdvertisementDao().findAll(),
```
데이터를 불러오는 것은 프레젠터가 할 일 아닐까요?

또한 왜 생성자로만 어댑터를 초기화 해야 할까요?
만약 불러와야 할 영화가 10000개라면 이 화면을 보여주기 위해서 네트워크 요청을 10분은 해야 할지도 모릅니다.

생성자로 데이터를 받는 방법이 아닌, notifyXXX() 형태로 업데이트 하도록 만들어주세요

hxeyexn replied:
데이터를 불러오는 일을 프레젠터에게 위임했습니다!
그리고 `notifyDataSetChanged` 를 이용해서 어댑터를 업데이트 하도록 변경했습니다! [f50de15](https://github.com/woowacourse/android-movie-theater/pull/55/commits/f50de159d224ff2c8f22a67b9b535baa32c4dce6)

laco-dev commented on app/src/main/java/woowacourse/movie/view/home/adapter/MovieCatalogAdapter.kt:
```diff
@@ -40,7 +41,8 @@ class MovieCatalogAdapter(
         when (CatalogViewType.from(getItemViewType(position))) {
             CatalogViewType.MOVIE -> {
-                val item = movies[position]
+                val moviePosition = position - (position / (CatalogViewType.ADVERTISEMENT.interval))
```
어려운 과제일 수 있습니다. 한번 고민해보고 너무 어렵다 싶으면 넘어가주세요

영화 사이에 몇번째에 광고가 보여야 한다는.
UI 로직일까요? 아니면 비즈니스 로직일까요?

hxeyexn replied:
광고의 위치에 따라 가장 크게 영향을 받는 건 수익이라고 생각했습니다. 수익을 얻을 수 있다는 건 비즈니스 로직에 가깝다는 생각이 드네요..! 그런데 또 한편으로는 광고가 눈에 보여진다만 생각해보면 UI 로직인 것 같기도 하고 어렵네요 ㅠㅠ 

광고가 보여지는 위치를 정하는 일은 보통 기획자가 주로 하는 일 같은 데, 기획자가 정하는 것들은 비즈니스 로직에 가깝다고 보면 될까요..?

laco-dev replied:
네, 기획자 == 비즈니스 라고 생각하는 것이 좋습니다.
반대로 말하면 우리가 작성하는 비즈니스 로직은 기획자도 알고 있어야 합니다.

그렇기 때문에 테스트 코드를 작성한다는 것을 프로그램의 기능 명세서를 작성한다고도 합니다.

laco-dev commented on app/src/main/java/woowacourse/movie/view/home/adapter/viewholder/AdvertisementViewHolder.kt:
```diff
@@ -1,15 +1,17 @@
+    private val binding: ItemAdvertisementBinding,
+) : RecyclerView.ViewHolder(binding.root) {
+    init {
+        binding.advertisement
```
???

hxeyexn replied:
헉 값 할당하는 걸 빼먹었네요.. DataBinding 다시 적용하면서 해당 부분도 수정해뒀습니다! [e222e2f](https://github.com/woowacourse/android-movie-theater/pull/55/commits/e222e2f997c8d001014466395c5e9839e907685e)

laco-dev commented on app/src/main/java/woowacourse/movie/view/home/adapter/viewholder/MovieViewHolder.kt:
```diff
@@ -1,29 +1,27 @@
+            itemMovieCatalogImageViewPoster.setImageResource(item.posterId)
+            itemMovieCatalogTextViewScreeningDate.text = convertPeriodFormat(item.screeningPeriod)
+            itemMovieCatalogTextViewRunningTime.text = item.runningTime
+            itemMovieCatalogButtonReservation.setOnClickListener { movie(item) }
```
movie() 라는 함수가 클릭 헀을 때 콜백으로 전달받는 것이라고 알 수 있을까요?
또한 이미 item으로 Movie를 전달하고 있다면, 굳이 다시 파라미터로 넘겨줄 필요가 없어보이네요

```kotlin
holder.bind(item) { foo(item) }
```

hxeyexn replied:
`onMovieSelected` 라는 이름으로 콜백 변수명 변경했습니다! 
그리고 movie가 아닌 movieId를 넘기도록 수정했습니다. [9475870](https://github.com/woowacourse/android-movie-theater/pull/55/commits/9475870a8d83d4e629c221134388834440c07916)

laco-dev commented on app/src/main/java/woowacourse/movie/view/reservation/ReservationDetailActivity.kt:
```diff
@@ -6,94 +6,88 @@ import android.util.Log
-                numberOfTickets.text = presenter.ticket.count.toString()
+                bundle.getInt(HEAD_COUNT)
+            }.onSuccess { headCount ->
+                presenter.headCount.restore(headCount)
```
프레젠터도 결국 다시 만들어지게 될텐데 생성 시점에 복원을 하는 방법은 없을까요?

```kotlin
val savedHeadCount = restoreHeadCount()
val presenter = Presenter(savedHeadCount)
```

hxeyexn replied:
생성 시점에 HeadCount를 복원 하도록 변경했습니다! [86425ef](https://github.com/woowacourse/android-movie-theater/pull/55/commits/86425efcbd8c565595dcae99c6439b135f18f817)

laco-dev commented on app/src/main/java/woowacourse/movie/view/reservation/ReservationDetailActivity.kt:
```diff
@@ -158,33 +153,25 @@ class ReservationDetailActivity : AppCompatActivity(), ReservationDetailContract
-            presenter.decreaseTicketCount(presenter.ticket.count)
-        }
-    }
+    fun initializeMinusButton() = presenter.decreaseHeadCount(presenter.headCount.count)
```
이 함수들은 왜 public으로 만들어졌나요?
초기화 하는 행위를 다른 곳에서 접근할 이유가 있는 건가요?

hxeyexn replied:
DataBinding을 충분히 활용하지 못해서, XML에서 해당 함수를 직접 호출할 수 있도록 함수를 public으로 설정했었습니다... 😱
DataBinding 을 다시 적용하면서 해당 함수는 제거했습니다! [c4a7ded](https://github.com/woowacourse/android-movie-theater/pull/55/commits/c4a7ded9dc487523ecba264509e16927f2ba527e)

laco-dev commented on app/src/main/java/woowacourse/movie/view/reservation/SeatSelectionActivity.kt:
```diff
@@ -99,7 +105,7 @@ class SeatSelectionActivity : AppCompatActivity(), SeatSelectionContract.View {
     ) {
         setOnClickListener {
-            if (getSeatsCount() < ticket.count || isSelected) {
+            if (getSeatsCount() < headCount.count || isSelected) {
```
뭔가 테스트 코드로 만들어질 만한 듯 해 보이는데요. 이 로직은 비즈니스 로직일까요?

> 선택한 좌석의 수가 전체 수량보다 작거나 선택되어 있다면 XXXX

hxeyexn replied:
해당 화면의 기능을 구현할 때 로직이 복잡해서 비즈니스 로직을 제대로 분리하지 못했던 것 같습니다 🥲 
위 로직을 presenter로 이동시켰습니다! [2521b09](https://github.com/woowacourse/android-movie-theater/pull/55/commits/2521b09158848f2ef12b093554431c3470a94e5c)

laco-dev commented on app/src/main/java/woowacourse/movie/view/reservation/SeatSelectionActivity.kt:
```diff
@@ -162,27 +167,51 @@ class SeatSelectionActivity : AppCompatActivity(), SeatSelectionContract.View {
+            showErrorToast()
+            finish()
+        }
+    }
```
전부다 이런 패턴으로 만들어 주셨는데, 앞서 남겨드린 의견 잘 참고해주세요.
지금 구조가 좋은 방법은 아닙니다.

hxeyexn replied:
이전 화면에서 정보를 전달받지 못했을 때 현재 화면을 유지한 채로 스낵바를 이용해 사용자에게 에러 문구(ex. `예기치 못한 오류가 발생해 이전 데이터를 불러오지 못했습니다. 이전 화면으로 돌아가시겠습니까?`)를 띄우고, 사용자가 확인 버튼을 눌렀을 때만 이전 화면으로 이동해 예매를 계속 진행할 수 있도록 변경했습니다! [d8e2875](https://github.com/woowacourse/android-movie-theater/pull/55/commits/d8e287517a3e1b61355cfcd2289104012eaad7d2)

laco-dev commented on app/src/main/java/woowacourse/movie/view/theater/TheaterSelectionFragment.kt:
```diff
@@ -0,0 +1,86 @@
+            }
+            else -> id
+        }
+    }
```
이 함수도 같은 맥락에서 동일한 의견입니다.

hxeyexn replied:
https://github.com/woowacourse/android-movie-theater/pull/55#discussion_r1590665029 해당 코멘트와 동일한 방식으로 변경했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/view/theater/adapter/TheaterSelectionAdapter.kt:
```diff
@@ -0,0 +1,32 @@
+
+class TheaterSelectionAdapter(
+    private val theaters: List<Theater>,
+    private val theaterId: OnSelectTheater,
```
마찬가지로 theaterId가 콜백이라는 것으로 보이지 않습니다.

hxeyexn replied:
onTheaterSelected 라는 이름으로 변경했습니다! [2889724](https://github.com/woowacourse/android-movie-theater/pull/55/commits/28897248c60b61dfc77e2e566cf25734f2d76866)

콜백 함수의 동작이 `극장을 선택했을 때 예매 화면으로 이동한다.` 라고 했을 때 `극장을 선택했을 때`와 같이 수행되어야 하는 동작에 초점을 맞추는 게 좋을지, 아니면 `예매 화면으로 이동한다.`와 같은 결과에 초점을 맞추는 게 좋을 지 고민되었습니다..! 

페로로는 보통 어떤 방식으로 콜백 변수명을 정하시는 지 궁금합니다!

laco-dev replied:
결과는 그 리스너를 구현하는 곳의 책임이니까요
인터페이스를 만드는 이유도 실제 구현체가 무엇을 하는지 결과를 알 필요가 없기 때문인데요.

명확한 이벤트를 요구하는 것이라면 화면 이동이라고 지을 수 있지만. 어댑터 구조에서는 실제로 화면을 이동할지 토스트를 띄울지는 구현하는 곳의 책임으로 볼 수 있습니다.

hxeyexn replied:
콜백 변수명은 동작에 초점을 맞추는 것이 더 좋겠네요 ㅎㅎ 페로로 피드백 덕분에 콜백 변수 네이밍에 대한 기준이 명확하게 생겼습니다! 감사합니다 😊

laco-dev commented on app/src/main/res/layout/activity_reservation_finished.xml:
```diff
@@ -1,138 +1,164 @@
+
+    <data>
+
+        <variable
+            name="reservation_finished"
+            type="woowacourse.movie.view.finished.ReservationFinishedActivity" />
```
실제 사용되고 있지 않네요.

hxeyexn replied:
제가 DataBinding을 ViewBinding 처럼 사용하고 있었던 것 같네요..!

> MVP 대원칙에 따라 프레젠터의 상태에 직접 접근하지 않습니다.
> Presenter로 클릭 동작과 같은 이벤트를 전달하고, 프레젠터에서 View에게 다시 업데이트를 요청하게 됩니다.
> 
> 이후 View가 데이터바인딩의 변수를 다시 할당함으로써 업데이트 되도록 하는 순서가 이어집니다.

DataBinding 관련해서 위에 코멘트 주신 내용 참고해 전반적으로 다 수정했습니다!

`홈 화면 DataBinding 적용` : [b8f6bf3](https://github.com/woowacourse/android-movie-theater/pull/55/commits/b8f6bf3f05f0f673bf89e752087233a25f2b1c04) , [e222e2f](https://github.com/woowacourse/android-movie-theater/pull/55/commits/e222e2f997c8d001014466395c5e9839e907685e)
`극장 선택 DataBinding 적용` : [521a3fa](https://github.com/woowacourse/android-movie-theater/pull/55/commits/521a3fa8cc79d7acf2654881cbc806a57836a153)
`예약 화면 DataBinding 적용` : [033daf6](https://github.com/woowacourse/android-movie-theater/pull/55/commits/033daf60e8bb5d3262d1cd6f55d9d2c8ceb452d3) , [c4a7ded](https://github.com/woowacourse/android-movie-theater/pull/55/commits/c4a7ded9dc487523ecba264509e16927f2ba527e)
`좌석 선택 화면 DataBinding 적용` : [05ceea9](https://github.com/woowacourse/android-movie-theater/pull/55/commits/05ceea936401c27449da3b675256c3448aba6270)
`예매 완료 화면 DataBinding 적용` : [598ce00](https://github.com/woowacourse/android-movie-theater/pull/55/commits/598ce0097bc3ada40402503ccc8877b6054c46ad)

laco-dev reviewed (changes requested):
> 데이터 바인딩 시 프레젠터 안에 있는 모델을 private를 풀어야 하는데, 이런 방식으로 바인딩을 진행하면 안전성이 떨어지는 것인가요? 바인딩을 처음 사용해봐서 궁금증이 생겼습니다.

데이터바인딩이 실제로 사용되고 있지 않아서, 답변을 드리기 어려운데요.
MVP 대원칙에 따라 프레젠터의 상태에 직접 접근하지 않습니다.
Presenter로 클릭 동작과 같은 이벤트를 전달하고, 프레젠터에서 View에게 다시 업데이트를 요청하게 됩니다.

이후 View가 데이터바인딩의 변수를 다시 할당함으로써 업데이트 되도록 하는 순서가 이어집니다.

hxeyexn committed:
refactor: HomeFragment test를 위한 fragment 실행 방법 변경

- 이전: MainActivity를 이용해 HomeFragment를 실행
- 이후: launchFragmentInContainer를 이용해 HomeFragment를 실행

hxeyexn committed:
refactor: 좌석 선택 fragment 테스트를 위한 fragment 실행 방법 변경

- 이전: MainActivity를 이용해 SeatSelectionFragment를 실행
- 이후: launchFragmentInContainer를 이용해 SeatSelectionFragment를 실행

hxeyexn committed:
refactor: 패키지 구조 변경

- 이전: 컴포넌트별
- 이후: 기능별

hxeyexn committed:
refactor: 사용하지 않는 fragment 이동 기능 삭제

hxeyexn committed:
refactor: 상영 시간 갯수 format 방식 변경

hxeyexn committed:
refactor: 예매 내역 복구 로직 변경

- 이전: HeadCount의 count 상태를 복구하는 restore 메서드 이용
- 이후: HeadCount 객체를 다시 생성

hxeyexn committed:
feat(ReservationDetail): 인원 수 복구 로직 변경

- 이전: HeadCount의 복구 로직 이용
- 이후: Presenter에서 HeadCount 새로 생성

hxeyexn committed:
test(HeadCount): 테스트명 수정 및 시나리오 변경

- 티켓 -> 인원 수로 변경
- 100장일 때 Failure 반환 -> 20장일 때 Failure 반환 검증 테스트로 변경

hxeyexn committed:
feat(ReservationDetail): 인원 수 증감 로직 변경

hxeyexn committed:
feat(ReservationDetail): 영화 ID를 프레젠터 주 생성자로 받도록 변경

hxeyexn committed:
feat(SeatSelection): 영화 ID를 프레젠터 주 생성자로 받도록 변경

hxeyexn committed:
refactor(TheaterSelection): 프레젠터 생성 로직 메서드로 분리

hxeyexn committed:
feat(ReservationDetail): 극장 ID를 프레젠터 주 생성자로 받도록 변경

hxeyexn committed:
feat(SeatSelection): 극장 ID를 프레젠터 주 생성자로 받도록 변경

hxeyexn committed:
refactor(Main): Bottom Menu 이동 로직 Util에서 액티비티로 이동

hxeyexn committed:
refactor(TheaterSelector): 콜백 함수 네이밍 변경

- 이전: theaterId
- 이후: onTheaterSelected

hxeyexn committed:
fix(TheaterSelection): 영화에 맞는 극장을 찾지 못하는 문제 해결

hxeyexn committed:
feat(Home): 극장 선택 화면으로 이동 시 movieId를 넘기도록 수정

- movie 대신 movieId를 넘기도록 수정
- 콜백 함수 네이밍 변경

hxeyexn committed:
feat(Ticket): 극장 Id가 아닌 극장 이름을 가지고 있도록 변경

hxeyexn committed:
refactor(ReservationDetail): 좌석 선택으로 티켓 전달하는 메서드명 변경

- UI와 직접적인 연관이 없도록 메서드명 변경

hxeyexn committed:
refactor(TheaterSelection): 예약 상세로 극장 전달하는 메서드명 변경

- UI와 직접적인 연관이 없도록 메서드명 변경

hxeyexn committed:
refactor(ReservationDetail): Reservation으로 네이밍 변경

hxeyexn committed:
feat(Reservation): 인원 수를 프레젠터 생성 시점에 복원되도록 변경

- 화면 회전 시 인원 수를 프레젠터 생성 시점에 복원되도록 변경
- ReservationPresenter의 HeadCount 상태를 private로 변경

hxeyexn committed:
feat(Reservation): Activity의 영화 ID 및 극장 ID 상태 삭제

- ReservationPresenter가 가지고 있도록 변경

hxeyexn committed:
feat(Reservation): DataBinding 재적용

- +/- 버튼, 좌석 선택 버튼, 인원 수 DataBinding 재적용

hxeyexn committed:
feat(Reservation): 영화 정보 DataBinding 적용

hxeyexn committed:
feat(SeatSelection): DataBinding 적용

hxeyexn committed:
feat(ReservationFinished): DataBinding 적용

hxeyexn committed:
refactor(SeatSelection): UI와 관계없이 기능 관점으로 메서드명 변경

- 이전: initializeConfirmButton
- 이후: requestReservationConfirm

hxeyexn committed:
feat(TheaterSelection): 극장 DB 구조 변경

- 하나의 시간과 하나의 영화를 가지고 있는 ScreeningSlot 객체 구현
- 이전: 극장에 하나의 영화와 하나의 상영시간을 고정
- 이후: 하나의 극장에서 여러 개의 영화를 상영할 수 있도록

hxeyexn committed:
refactor(TheaterSelection): 극장 정보 전달 메서드명 변경

- ReservationDetail을 Reservation으로 변경함에 따라, 관련 이동 메서드의 이름도 수정

hxeyexn committed:
style: xml 코드 포맷팅

hxeyexn committed:
refactor(Movie): DataBinding 재적용 및 id 수정

- MovieViewHolder DataBinding 재적용
- xml view Id 수정

hxeyexn committed:
refactor(Advertisement): DataBinding 재적용 및 id 수정

- AdvertisementViewHolder DataBinding 재적용
- xml view Id 수정

hxeyexn committed:
feat(Home): 어댑터 초기화 방법 및 영화, 광고 데이터 로딩 방법 변경

- 이전: activity에서 데이터를 로딩해 이를 어댑터의 생성자로 받아 어댑터 초기화
- 이후: presenter가 데이터를 로딩, notifyDataSetChanged를 이용해 어댑터 갱신하도록 변경

hxeyexn committed:
feat(TheaterSelection): 어댑터 초기화 방법 및 극장 데이터 로딩 방법 변경

- 이전: activity에서 데이터를 로딩해 이를 어댑터의 생성자로 받아 어댑터 초기화
- 이후: presenter가 데이터를 로딩, notifyDataSetChanged를 이용해 어댑터 갱신하도록 변경

hxeyexn committed:
refactor(TheaterSelection): DataBinding 재적용 및 id 수정

- TheaterViewHolder DataBinding 재적용
- xml view Id 수정

hxeyexn committed:
feat(SeatSelction): 확인 버튼 Enabled 여부 결정 로직 수정

hxeyexn committed:
feat(SeatSelection): 예약 가능 판단 로직 변경, 프레젠터 상태 변경

- 예약 가능 판단 메서드(확인 버튼 enabled 여부 판단)의 시그니처 int 에서 boolean으로 변경
- SeatSelection의 Seats 상태를 private으로 변경

hxeyexn committed:
feat(SeatSelection): 에러 핸들링 방법 변경

영화 ID, 극장 ID, 인원 수, 상영 시간 중 하나라도 값을 전달받지 못했을 때 에러 핸들링
- 이전: 에러 toast 보여준 후 좌석 선택 화면 종료
- 이후: 에러 snack bar 보여준 후 사용자가 확인 버튼을 눌렀을 때 이전 화면으로 돌아가도록 변경

hxeyexn committed:
refactor(TheaterSelection): 불필요한 캐스팅 삭제

hxeyexn committed:
feat(Reservation): 에러 핸들링

영화 ID, 극장 ID 중 하나라도 값을 전달받지 못했을 때 에러 snack bar 보여준 후 사용자가 확인 버튼을 눌렀을 때 이전 화면으로 돌아가도록 변경

hxeyexn committed:
feat(Reservation): 에러 핸들링 및 티켓을 프레젠터 생성자로 주입

- 티켓을 프레젠터 생성자로 주입
- 티켓을 전달받지 못했을 때 에러 snack bar 보여준 후 사용자가 홈으로 이동 버튼을 눌렀을 때 홈 화면으로 돌아가도록 변경

hxeyexn committed:
refactor(SeatSelection): seats 에러 핸들링 방법 변경

hxeyexn committed:
feat(TheaterSelection): 영화 ID 에러 핸들링 방법 변경

- 이전: 에러 toast 보여준 후 좌석 선택 화면 종료
- 이후: 에러 snack bar 보여준 후 사용자가 확인 버튼을 눌렀을 때 이전 화면으로 돌아가도록 변경

hxeyexn committed:
refactor(Reservation): 인원 수 범위 핸들링 로직 접근 제한자 변경

hxeyexn committed:
feat(SeatSelction): 비즈니스 로직 presenter로 이동

- 좌석 선택 시 좌석 선택 상태, 현재 금액, 확인 버튼 상태 업데이트 로직 presenter로 이동

hxeyexn committed:
refactor: xml id 축약형으로 수정

hxeyexn committed:
refactor(Home): 프레젠터, 계약 네이밍 프래그먼트와 동일하게 변경

hxeyexn commented:
안녕하세요 페로로! 
꼼꼼히 피드백 남겨 주셔서 정말 많은 도움이 되었습니다 🙇 남겨주신 모든 피드백에 답변 달아두었습니다 😊
[seats와 amount 관계](https://github.com/woowacourse/android-movie-theater/pull/55#discussion_r1590334509) , [광고의 주기](https://github.com/woowacourse/android-movie-theater/pull/55#discussion_r1590324024) 관련된 코멘트를 제외한 모든 피드백 반영했습니다! 

[seats와 amount 관계](https://github.com/woowacourse/android-movie-theater/pull/55#discussion_r1590334509) , [광고의 주기](https://github.com/woowacourse/android-movie-theater/pull/55#discussion_r1590324024) 는 이해가 잘 되지 않아서 질문 남겨두었습니다..! 코멘트 주신다면 다음 리뷰 요청 때 반영해보겠습니다!

추가적으로 질문 남겨둔 코멘트인데, 코멘트가 많아 찾기 힘드실 것 같아서 아래에 링크 연결해뒀습니다!
[확신이 서지 않는 코멘트](https://github.com/woowacourse/android-movie-theater/pull/55#discussion_r1590692807)
[콜백 변수 네이밍 질문](https://github.com/woowacourse/android-movie-theater/pull/55#discussion_r1590327903)

laco-dev commented on app/build.gradle.kts:
```diff
@@ -57,6 +57,7 @@ dependencies {
     testImplementation("org.assertj:assertj-core:3.25.3")
     testImplementation("io.kotest:kotest-runner-junit5:5.8.0")
     testImplementation("io.mockk:mockk:1.9.3")
+    debugImplementation("androidx.fragment:fragment-testing:$1.7.0")
```
`androidx.fragment:fragment-testing:$1.7.0` 어떤 코드를 참고하셨는지 모르겠는데, $는 변수를 참조하기 위해 사용하는 값이에요.
하드코딩이라면 사용하지 않습니다.

hxeyexn replied:
```kotlin
dependencies {
    val fragment_version = "1.7.0"

    debugImplementation("androidx.fragment:fragment-testing:$fragment_version")
}
```

공식문서에 위 코드를 참고하다가 $ 삭제를 안 했네요...! $ 제거하도록 하겠습니다!

laco-dev commented on app/src/main/AndroidManifest.xml:
```diff
@@ -22,16 +22,16 @@
         <activity
-            android:name=".view.finished.ReservationFinishedActivity"
-            android:parentActivityName=".view.MainActivity"
+            android:name=".feature.finished.ReservationFinishedActivity"
```
`feature` 패키지 분리 좋습니다. 👍🏻

laco-dev commented on app/src/main/java/woowacourse/movie/feature/finished/ReservationFinishedActivity.kt:
```diff
@@ -0,0 +1,87 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding.activity = this
+        handleBackPressed()
+        initPresenter()
```
제가 보기에는 둘다 초기화 하는 작업으로 보이는데 handle()과 init()은 어떤 차이가 있을까요?

hxeyexn replied:
`handle`은 사용자의 액션을 처리(ex. 뒤로 가기 버튼을 눌렀을 때 이전 화면으로 돌아간다)하는 메서드이고, `init`은 단지 객체를 초기화(ex. Presenter를 사용할 수 있도록 초기화)한다는 차이점이 있습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/feature/finished/ReservationFinishedActivity.kt:
```diff
@@ -0,0 +1,87 @@
+        binding.activity = this
+        handleBackPressed()
+        initPresenter()
+        with(presenter) {
+            handleUndeliveredTicket()
+            loadMovie()
+            loadTicket()
+        }
```
handle()에서 문제가 있더라도 loadMovie() loadTicket()은 정상적으로 호출이 되지 않을까요?
프레젠터의 역할로 본다면 아래와 같은 로직이 되어야 하지 않을까요?
```kotlin
// Presenter
val movieId = ...
if (isValidMovie(movieId)) { 
  load()
  ...
} else {
   handleUndeliveredTicket()
}
```

hxeyexn replied:
아래와 같은 로직으로 변경했습니다~!
```kotlin   
    override fun loadMovieInformation() {
        if (isValidMovieInformation()) {
            loadMovie()
            loadScreeningDates()
            loadScreeningTimes()
        } else {
            handleUndeliveredData()
        }
    }
 ```

laco-dev commented on app/src/main/java/woowacourse/movie/feature/finished/ReservationFinishedActivity.kt:
```diff
@@ -0,0 +1,87 @@
+            val seats = ticket.seats.seats
+            headCount = seats.size
+            amount = convertAmountFormat(this@ReservationFinishedActivity, ticket.amount)
+            this.seats = seats.joinToString(getString(R.string.reservation_finished_seat_separator)) { "${it.row}${it.column}" }
```
이 값이 실제로는 한번만 사용되는데 변수로 만들필요까지 있을까? 라는 생각은 드네요
다만 연습하는 과정이니 크게 중요한 요소는 아닙니다.
```kotlin
binding.tvSeats.text = ... // 와 별로 차이가 없는 정도
```

hxeyexn replied:
미션을 진행하면서 데이터 바인딩을 변경점이 있는 값만 변수로 만들지, 모든 값을 변수로 만들지 고민 했었습니다! 최대한 많은 곳에 사용해보면서 장단점을 느껴보고 싶어, 이번 미션에선 최대한 모든 값을 변수로 만들어 보았습니다. 위와 같이 코드를 작성하면서 페로로가 말씀해주신 것 처럼 한번만 사용 되는 값들을 변수로 만들었을 때 데이터 바인딩의 장점이 크게 부각되지는 않는다고 생각했습니다 🙂 한번만 사용되는 값들은 view binding만 해도 충분할 것 같습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/feature/finished/ReservationFinishedActivity.kt:
```diff
@@ -0,0 +1,87 @@
+
+    override fun navigateToHome() {
+        val intent = Intent(this@ReservationFinishedActivity, MainActivity::class.java)
+        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
```
addFlags는 Intent 자신을 반환하기 때문에 `Intent(...).addFlags()` 처럼 작성할 수 있습니다.

hxeyexn replied:
감사합니다! 해당 부분 아래와 같이 변경 했습니다~

``` kotlin
val intent =
    Intent(this@ReservationFinishedActivity, MainActivity::class.java)
        .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
```

laco-dev commented on app/src/main/java/woowacourse/movie/feature/home/HomeFragment.kt:
```diff
@@ -35,11 +35,23 @@ class HomeFragment : Fragment(), ReservationHomeContract.View {
+        movies: List<Movie>,
+        advertisements: List<Advertisement>,
+    ) {
+        val movieCatalogAdapter =
+            MovieCatalogAdapter { movieId ->
+                presenter.sendMovieIdToTheaterSelection(movieId)
+            }
```
어댑터 초기화를 show 함수에서 할 필요가 있나요?
가령 show 함수가 여러번 호출되면 매번 새로운 어댑터가 할당 됩니다.

이 함수에서는 updateData()만 호출해도 문제가 없어 보여요

hxeyexn replied:
해당 함수에서 updateData()만 호출하도록 수정했습니다!

```kotlin
    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        ...
        initMovieCatalogAdapter()
        ...
    }

    override fun showMovieCatalog(
        movies: List<Movie>,
        advertisements: List<Advertisement>,
    ) {
        movieCatalogAdapter.updateData(movies, advertisements)
    }
    
     private fun initMovieCatalogAdapter() {
        movieCatalogAdapter =
            MovieCatalogAdapter { movieId ->
                presenter.sendMovieIdToTheaterSelection(movieId)
            }
        binding.rvHome.adapter = movieCatalogAdapter
    }

```

laco-dev commented on app/src/main/java/woowacourse/movie/feature/home/adapter/viewholder/AdvertisementViewHolder.kt:
```diff
@@ -8,10 +8,10 @@ class AdvertisementViewHolder(
 ) : RecyclerView.ViewHolder(binding.root) {
     init {
-        binding.advertisement
+        binding.viewHolder = this
```
데이터바인딩 변수로 뷰홀더를 넣어주는 이유가 있나요?
아무런 메소드가 없는데, 어떤 역할을 기대하는 건지 궁금해서요

hxeyexn replied:
데이터바인딩이 사용하는 파일에 대해서 변수를 선언하고, 해당 값을 바인딩 해줘야 데이터 바인딩이 활성화 된다고 잘못 알고 있었습니다...! 3, 4단계를 진행하면서 불필요한 바인딩 변수들은 제거했습니다 🥹

laco-dev commented on app/src/main/java/woowacourse/movie/feature/reservation/ReservationActivity.kt:
```diff
@@ -0,0 +1,198 @@
+        val savedHeadCount = bringSavedHeadCount(savedInstanceState)
+        initPresenter(savedHeadCount)
+
+        with(presenter) {
+            handleUndeliveredData()
+            loadMovie()
+            loadScreeningPeriod()
+        }
+        changeHeadCount(savedHeadCount)
+        updateScreeningTimes(DEFAULT_TIME_ID)
```
결과적으로 HeadCount나 상영 시간 등은 프레젠터에 의해 판단하도록 이루어 져야 합니다.

프레젠터는 프레젠터대로 초기화하고 뷰는 기본값으로 초기화 한다면, 사용자에게 데이터를 불러오기 전까지 알 수 없는 이상한 정보들이 출력됩니다.

프레젠터가 아이디가 없다 -> 기본 값을 보여준다
프레젠터가 아이디가 있다 -> 데이터를 불러와서 보여준다

즉 이 모든 일련의 로직은 프레젠터가 관리합니다.

hxeyexn replied:
MVP 구조를 제대로 적용하고 있지 못했던 것 같습니다 😭 이 코멘트 덕분에 MVP에 대한 이해가 한층 더 깊어진 것 같습니다! 감사합니다 🙇 프레젠터가 프레젠터와 뷰의 기본값을 초기화하도록 변경했습니다!

### presenter
``` kotlin
    init {
        view.changeHeadCount(savedHeadCount)
    }

    override fun loadMovieInformation() {
        if (isValidMovieInformation()) {
            loadMovie()
            loadScreeningDates()
            loadScreeningTimes()
        } else {
            handleUndeliveredData()
        }
    }

```

### activity
```kotlin
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ...
        presenter.loadMovieInformation()
        ...
    }
```

laco-dev commented on app/src/main/java/woowacourse/movie/feature/reservation/ReservationActivity.kt:
```diff
@@ -0,0 +1,198 @@
+        count: HeadCount,
+    ) {
+        val intent = Intent(this, SeatSelectionActivity::class.java)
+        intent.apply {
+            putExtra(MOVIE_ID, movieId)
+            putExtra(THEATER_ID, theaterId)
+            putExtra(SCREENING_DATE_TIME, dateTime)
+            putExtra(HEAD_COUNT, count)
+        }
```
putExtra도 Intent 자신을 반환하기 때문에 굳이 apply 를 사용하지 않아도 됩니다.

hxeyexn replied:
putExtra도 연쇄적으로 호출이 가능하군요! 아래와 같이 수정했습니다 😊
```kotlin
        val intent =
            Intent(this, SeatSelectionActivity::class.java)
                .putExtra(MOVIE_ID, movieId)
                .putExtra(THEATER_ID, theaterId)
                .putExtra(SCREENING_DATE_TIME, dateTime)
                .putExtra(HEAD_COUNT, count)
```

laco-dev commented on app/src/main/java/woowacourse/movie/feature/reservation/ReservationPresenter.kt:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    override fun sendTicketToSeatSelection() {
+        val date: String = view.getScreeningDate()
+        val time: String = view.getScreeningTime()
```
View도 프레젠터와 마찬가지로 값을 반환하지 않아요.
수동적으로 View의 값을 꺼내오는 것이 아닌 능동적으로 프레젠터에게 값을 넘겨주도록 합니다.

다른 곳에서는 다 잘 작성 해주셨는데 왜 이곳에서는 꺼내와야지 라는 것으로 도달했는지 궁금합니다.

hxeyexn replied:
원래 상영일과 상영시간이 변경될 때마다 프레젠터에서 해당 값을 갱신하는 로직이 아니라, 좌석 선택 버튼을 클릭 했을 때 스피너의 선택된 아이템을 확인하는 방식이었습니다. 좌석 선택 버튼을 눌러야만 값을 얻을 수 있는 데 이를 어떻게 프레젠터로 넣어줘야하나 고민하다가 꺼내와야겠다는 생각에 도달했던 것 같습니다. 지금은 프레젠터가 date id와 time id를 가지고 있고, 스피너의 item이 선택될 때마다 해당 값들을 갱신해주는 로직으로 변경했습니다!

```kotlin
class ReservationPresenter(
    ...
) {
    ...
    private var screeningDateId: Long = 0
    private var screeningTimeId: Long = 0
    ...

    override fun sendReservationInformationToSeatSelection() {
        val screeningDate = screeningDates[screeningDateId.toInt()]
        val screeningTime = screeningTimes[screeningTimeId.toInt()]
        val dateTime = LocalDateTime.of(screeningDate, screeningTime)
        view.showDateTime(dateTime)
        view.navigateToSeatSelection(dateTime, movieId, theaterId, headCount)
    }
    ...
}
```

laco-dev commented on app/src/main/java/woowacourse/movie/feature/reservation/ReservationContract.kt:
```diff
@@ -0,0 +1,56 @@
+            selectedDate: String,
+        )
+
+        fun changeHeadCount(count: Int)
```
show()와 change()는 차이가 있나요?

hxeyexn replied:
> show: 로드된 데이터 값을 보여줄 때
> change: 사용자가 view와 상호작용해서(ex. + 버튼을 눌렀을 때 인원 수가 증가) 변하는 값을 보여줄 때

위와 같은 기준으로 show와 change를 사용했습니다! 페로로가 느끼시기엔 기준이 조금 모호한가용..?

laco-dev commented on app/src/main/java/woowacourse/movie/feature/reservation/ReservationPresenter.kt:
```diff
@@ -0,0 +1,76 @@
+        view.navigateToSeatSelection(dateTime, movieId, theaterId, headCount)
+    }
+
+    override fun restoreHeadCount() {
+        view.changeHeadCount(headCount.count)
+    }
```
프레젠터가 생성되면서 init 블록에서 호출하는 것과 차이가 있나요?

hxeyexn replied:
init 블록에서 호출하는 것과 차이가 없겠네요..! 해당 메서드를 init 에서 호출하도록 수정했습니다~
```kotlin
    init {
        view.changeHeadCount(savedHeadCount)
    }
```

laco-dev commented on app/src/main/java/woowacourse/movie/feature/seatselection/SeatSelectionActivity.kt:
```diff
@@ -163,55 +155,62 @@ class SeatSelectionActivity : AppCompatActivity(), SeatSelectionContract.View {
+                SeatsDao(),
+                ScreeningDao(),
+                TheaterDao(),
+                receiveMovieId(),
+                receiveTheaterId(),
+                headCount,
+                screeningDateTime,
```
Extra로 넘겨주는 파라미터 정보가 많으면, 별도 객체로 포장해서 주고 받는 방법도 사용할 수 있습니다.

hxeyexn replied:
단지 데이터 전달용으로 객체를 만들어도 될까 고민 했었는 데, 넘겨줄 파라미터 정보가 많을 때는 객체로 포장하는 게 확실히 가독성이 더 좋겠네요! 😊 아직 코드에 반영하지는 못했습니다..! ㅠㅠ

laco-dev commented on app/src/main/java/woowacourse/movie/feature/seatselection/SeatSelectionContract.kt:
```diff
@@ -34,10 +34,18 @@ interface SeatSelectionContract {
+        fun handleUndeliveredData()
+
+        fun updateReservationState(
+            seat: Seat,
+            index: Int,
```
해당 좌석이 몇번째 위치인지는 프레젠터가 가지고 있는 좌석들로부터 판단할 수 있지 않을까요?

hxeyexn replied:
좌석의 위치를 프레젠터가 가지고 있는 좌석들로 판단하도록 변경했습니다! 😊

laco-dev commented on app/src/main/java/woowacourse/movie/feature/seatselection/SeatSelectionPresenter.kt:
```diff
@@ -39,47 +43,73 @@ class SeatSelectionPresenter(
+    }
+
+    override fun handleUndeliveredData() {
+        if (movieId == -1 ||
+            theaterId == -1 ||
+            headCount.count == 0 ||
+            screeningDateTime == ScreeningDateTime("", "")
+        ) {
+            view.showErrorSnackBar()
+        }
```
임의의 값을 유효하지 않은 것으로 취급 하는 것도 늘 좋은 방법은 아니에요.

Int 값이 잘못되면 -1이 아니고 -2가 들어올수도있고 0이 잘못된 걸수도 있구요.
DateTime에 String이 empty가 아니고 "error" 이렇게 들어올지 아무도 모르는 거구요.

상수로 만들거나 하는 방법을 사용할 수 있지만 궁극적으로 유효하지 않은 값을 임의로 선정하는 것보다 오히려 null로 처리하는게 훨씬 직관적일 수 있습니다.

*읽고보니 왜 상영 시간을 String으로 출력하는지 모르겠네요.
왜인가요? UI에서 출력 방식 바꾸면 도메인 모델을 수정하나요?

hxeyexn replied:
유효하지 않은 값들도 비논리 오류라 null로 처리하는 게 더 직관적이겠네요!! headCount, screeningDateTime null로 변경했습니다! 
```kotlin
private fun receiveMovieId() =
    intent.getIntExtra(
        MOVIE_ID,
        DEFAULT_MOVIE_ID,
    )
```
다만 movieId와 theaterId는 위와 같이 getIntExtra를 이용하면 디폴트 값을 설정해야하고, 반환값이 널을 허용하지 않아서 우선 기본값을 상수로 만들어서 사용했습니다 ㅠㅠ getIntExtra도 값이 제대로 들어오지 않았을 때 null로 처리할 수 있는 방법이 있는 지 궁금합니다!

UI에서 출력할 때만 사용된다고 생각 했는 데 다시 서버에 전송해야한다는 걸 놓쳤습니다..! 3, 4단계 미션을 진행하면서 상영 시간 String을 LocalDate와 LocalTime으로 타입 변경했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/feature/theater/adapter/TheaterSelectionViewHolder.kt:
```diff
@@ -0,0 +1,27 @@
+        onTheaterSelected: OnTheaterSelected,
+    ) {
+        with(binding) {
+            this.theater = theater
+            this.screeningCount = screeningCount
```
지금 당장 문제가 되 보이지는 않지만 바인딩 설정 이후에 UI 업데이트가 이루어지지 않는다면, binding.executePendingBindings()을 참고해보세요

laco-dev commented on app/src/main/res/layout/activity_reservation.xml:
```diff
@@ -1,25 +1,49 @@
+            name="head_count"
+            type="int" />
+
+        <variable
+            name="date_time"
+            type="woowacourse.movie.model.movie.ScreeningDateTime" />
```
데이터바인딩에 사용하는 변수는 일반적으로 카멜 케이스를 지향하고 있어요.

hxeyexn replied:
xml id가 스네이크 케이스를 지향해서 데이터 바인딩에 사용하는 변수도 동일할 거라고 생각했네요 ㅠ 😱 감사합니다!

laco-dev commented on app/src/main/res/layout/item_movie_catalog.xml:
```diff
@@ -1,70 +1,85 @@
-        app:layout_constraintStart_toStartOf="@id/item_movie_catalog_text_view_title"
-        app:layout_constraintTop_toBottomOf="@id/item_movie_catalog_text_view_screening_date"
-        tools:text="러닝타임: 152분" />
+        <androidx.appcompat.widget.AppCompatImageView
```
`<androidx.appcompat.widget.AppCompatImageView` 와 `<ImageView` 는 어떤 차이가 있나요?

hxeyexn replied:
`ImageView`와 달리 `AppCompatImageView`는 오래된 버전의 안드로이드에서도 호환성이 좋다고 알고 있습니다. 또한 `AppCompatImageView`는 `ImageView`보다 사용할 수 있는 속성이 더 많아 기능적으로 더 좋습니다!

laco-dev reviewed (approved):
뷰가 해야할 일을 프레젠터가 하고 있고, 프레젠터가 해야할 일을 뷰가 하고 있는지 다시 검토해 볼 필요가 있습니다.

데이터바인딩 이전에 MVP 패턴의 기본적인 역할에 맡도록이요.
다음 미션에는 안드로이드 컴포넌트를 더 활용하는 연습을 하기에 도메인 수정은 깊게 다루지 않으셔도 됩니다.

질문에 대한 답변 포함해서 남겨드린 의견은 다음 미션에서 같이 반영 해주세요
